### PR TITLE
DAOS-9998 daos: add daos interoperability function

### PR DESCRIPTION
**Pull request by wangshilong**
1. add 'dmg pool upgrade command' interface
 2. versioning scheme with new associated pool properties
 3. restart after shutdown/failure
 4. new container created in a pool with a old version should not enable the new features.
 5. add new pool/container properties added between 2.0 and 2.2
 6. downgrading the DAOS software of an upgraded pool report a proper RAS error.
 7. revoke open handle and make sure there is no connection while pool upgraing.

Todo:
check that no rebuild/reint/extend in progress

Signed-off-by: Wang Shilong <shilong.wang@intel.com>

### CODE REVIEW

File: src/control/lib/control/pool.go [hunk 1]
```diff
@@ -588,7 +618,7 @@ func PoolGetProp(ctx context.Context, rpcClient UnaryInvoker, req *PoolGetPropRe
 	for _, prop := range resp {
 		pbProp, found := pbMap[prop.Number]
 		if !found {
-			return nil, errors.Errorf("unable to find prop %d (%s) in resp", prop.Number, prop.Name)
+			continue
```
[the remark this answers is not in the document]
Comment (wangshilong) [reply src/control/lib/control/pool.go:621]: For upgrading cases, eg for 2.0 to 2.2, there are a couple of properties are missed for old versions(which was expected).

File: src/container/srv_layout.h [hunk 2]
```diff
@@ -48,6 +48,7 @@ extern d_iov_t ds_cont_prop_version;		/* uint32_t */
 extern d_iov_t ds_cont_prop_cuuids;		/* container UUIDs KVS */
 extern d_iov_t ds_cont_prop_conts;		/* container KVS */
 extern d_iov_t ds_cont_prop_cont_handles;	/* container handle KVS */
+extern d_iov_t ds_cont_prop_global_version;	/* uint32 */
```
Comment (liw) [inline src/container/srv_layout.h:51]: Note that this is per-pool, not per-container. If the intention is to have a per-container property, we need to add to the Container Property KVS below. If the intention is to have a per-pool property, we can just export ds_pool_prop_global_version via some ds_pool API or field, without any new keys here.

**Review by liw — COMMENTED** [review]
Some comments I have so far; not finished yet.

File: src/pool/srv_pool.c [hunk 3]
```diff
@@ -3600,6 +3723,314 @@ ds_pool_prop_set_handler(crt_rpc_t *rpc)
...
+	rdb_tx_end(&tx);
+out_svc:
+	if (need_put_leader) {
+		ds_rsvc_set_hint(&svc->ps_rsvc, po_hint);
+		pool_svc_put_leader(svc);
+	}
+
+	if (upgraded) {
+		rc1 = ds_cont_upgrade(pool_uuid, svc->ps_cont_svc);
+		rc = ds_pool_mark_upgrade_completed(pool_uuid, svc, rc1);
```
Comment (liw) [inline src/pool/srv_pool.c:4008]: If both need_put_leader and upgraded are true, it looks like we are using svc without a reference.

File: src/pool/srv_pool.c [hunk 4]
```diff
@@ -3600,6 +3723,314 @@ ds_pool_prop_set_handler(crt_rpc_t *rpc)
 	crt_reply_send(rpc);
 }
 
+int pool_upgrade_props(struct rdb_tx *tx, struct pool_svc *svc, uuid_t pool_uuid,
+		       crt_rpc_t *rpc)
```
Comment (liw) [inline src/pool/srv_pool.c:3727]: [Nit] Static?

File: src/pool/srv_pool.c [hunk 5]
```diff
@@ -3600,6 +3723,314 @@ ds_pool_prop_set_handler(crt_rpc_t *rpc)
...
+		need_commit = true;
+	}
+
+	d_iov_set(&value, &val32, sizeof(val32));
+	rc = rdb_tx_lookup(tx, &svc->ps_root, &ds_pool_prop_global_version, &value);
+	if (rc && rc != -DER_NONEXIST) {
+		D_GOTO(out_free, rc);
+	} else if (rc == -DER_NONEXIST || val32 < DAOS_META_GLOBAL_VERSION) {
+		val32 = DAOS_META_GLOBAL_VERSION;
+		rc = rdb_tx_update(tx, &svc->ps_root, &ds_pool_prop_global_version, &value);
```
Comment (liw) [inline src/pool/srv_pool.c:3823]: Should we update ds_pool_prop_global_version only when the upgrade to GLOBAL_VERSION has completed? If we update it here, we no longer know what previous version is, especially after a system restart. Then how can we know what actions (i.e., the concept from https://daosio.atlassian.net/wiki/spaces/DC/pages/11092492289/Interoperability+up+down+grade) need to be taken in order to upgrade to GLOBAL_VERSION?

**Comment by wangshilong** [comment]
> Test stage checkpatch completed with status FAILURE. https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-8312/9/execution/node/112/log

Files are generated automatically, not sure how we could fix such kind of error messages.

10:19:14  2022-03-28 02:19:13,742 add_comment src/mgmt/pool.pb-c.c 1238 ERROR CODE_INDENT ' code indent should use tabs where possible'
10:19:14  2022-03-28 02:19:13,742 add_comment src/mgmt/pool.pb-c.c 1238 WARNING LEADING_SPACE ' please, no spaces at the start of a line'

File: src/container/srv_container.c [hunk 6]
```diff
@@ -2410,37 +2440,72 @@ cont_prop_read(struct rdb_tx *tx, struct cont *cont, uint64_t bits,
...
 			val = DAOS_PROP_PO_EC_PDA_DEFAULT;
-		} else  if (rc != 0) {
+		else if (rc != 0)
 			D_GOTO(out, rc);
-		}
 		D_ASSERT(idx < nr);
 		prop->dpp_entries[idx].dpe_type = DAOS_PROP_CO_EC_PDA;
 		prop->dpp_entries[idx].dpe_val = val;
+		if (rc == -DER_NONEXIST) {
+			prop->dpp_entries[idx].dpe_flags |= DAOS_PROP_ENTRY_NEGATIVE;
```
Comment (kjacque) [inline src/container/srv_container.c:2451]: Looking at the usage of this `DAOS_PROP_ENTRY_NEGATIVE` flag, might be better to name it something like `DAOS_PROP_ENTRY_NOT_SET`

File: src/container/srv_container.c [hunk 7]
```diff
@@ -3440,6 +3513,168 @@ ds_cont_list(uuid_t pool_uuid, struct daos_pool_cont_info **conts,
...
+	if (rc && rc != -DER_NONEXIST)
+		goto out;
+	/* latest container, nothing to update */
+	if (rc == 0 && global_ver == DS_POOL_GLOBAL_VERSION)
+		goto out;
+
+	if (global_ver > DS_POOL_GLOBAL_VERSION) {
+		D_ERROR("Downgrading pool/cont: "DF_CONTF" not supported\n",
+			DP_CONT(ap->pool_uuid, cont_uuid));
+		rc = -DER_INVAL;
```
Comment (kjacque) [inline src/container/srv_container.c:3570]: Might be worth adding a new DER error for this case (`DER_NOTSUPPORTED` maybe?)

File: src/container/srv_layout.c [hunk 8]
```diff
@@ -49,10 +49,77 @@ RDB_STRING_KEY(ds_cont_prop_, roots);
 RDB_STRING_KEY(ds_cont_prop_, ec_cell_sz);
 RDB_STRING_KEY(ds_cont_prop_, ec_pda);
 RDB_STRING_KEY(ds_cont_prop_, rp_pda);
+RDB_STRING_KEY(ds_cont_prop_, cont_global_version);
 
 /* dummy value for container roots, avoid malloc on demand */
 static struct daos_prop_co_roots dummy_roots;
 
+/** default properties, should cover all optional container properties */
+struct daos_prop_entry cont_prop_entries_default_v0[CONT_PROP_NUM_V0] = {
```
Comment (kjacque) [inline src/container/srv_layout.c:58]: I wouldn't block on this, but I wonder if there is a more elegant way to do this than duplicating these. With the ACL stuff below, it seems like we do some extra work to handle versions of this prop that hasn't actually changed between versions.

File: src/mgmt/srv_drpc.c [hunk 9]
```diff
@@ -990,55 +1046,65 @@ add_props_to_resp(daos_prop_t *prop, Mgmt__PoolGetPropResp *resp)
 {
 	Mgmt__PoolProperty	**resp_props;
 	struct daos_prop_entry	*entry;
-	int			 i, rc = 0;
+	int			 i, rc = 0, valid_prop_nr = 0, j = 0;
```
Comment (kjacque) [inline src/mgmt/srv_drpc.c:1049]: The DAOS style guide suggests putting each variable declaration on its own line. I know we don't always stick to that, but this line is getting long and includes, assignments, too. I think it would be best to break it up into individual lines.

File: src/mgmt/srv_drpc.c [hunk 10]
```diff
@@ -990,55 +1046,65 @@ add_props_to_resp(daos_prop_t *prop, Mgmt__PoolGetPropResp *resp)
...
 
 	if (prop == NULL || prop->dpp_nr == 0)
 		return 0;
 
-	D_ALLOC_ARRAY(resp_props, prop->dpp_nr);
+	for (i = 0; i < prop->dpp_nr; i++) {
+		entry = &prop->dpp_entries[i];
+		if (!(entry->dpe_flags & DAOS_PROP_ENTRY_NEGATIVE))
```
Comment (kjacque) [inline src/mgmt/srv_drpc.c:1056]: Maybe we should have an inline helper function called something like `prop_is_set` for checking this flag. Looks like we do this in a number of places in C code now.

File: src/pool/srv_pool.c [hunk 11]
```diff
@@ -255,6 +260,12 @@ pool_prop_default_copy(daos_prop_t *prop_def, daos_prop_t *prop)
 					return rc;
 			}
 			break;
+		case DAOS_PROP_PO_GLOBAL_VERSION:
+			D_ERROR("pool global version property could be not set\n");
+			return -DER_NO_PERM;
```
Comment (kjacque) [inline src/pool/srv_pool.c:265]: I would lean against -DER_NO_PERM as the return code here. -DER_NO_PERM is used to represent access denied via the ACL. I'd suggest either using -DER_INVAL or adding a new DER code (-DER_IMMUTABLE or something like that?)

File: src/pool/srv_pool.c [hunk 12]
```diff
@@ -3628,6 +3765,355 @@ ds_pool_prop_set_handler(crt_rpc_t *rpc)
...
+	d_iov_set(&value, &val, sizeof(val));
+	rc1 = rdb_tx_update(&tx, &svc->ps_root, &ds_pool_prop_upgrade_status,
+			   &value);
+	if (rc1)
+		D_GOTO(out_tx, rc1);
+
+	/* only bump global version if upgrade succeed. */
+	if (rc == 0) {
+		val = DS_POOL_GLOBAL_VERSION;
+		rc1 = rdb_tx_update(&tx, &svc->ps_root,
```
Comment (kjacque) [inline src/pool/srv_pool.c:3942]: Should be updating `value` with the contents of `val` in this case, yes?

File: src/pool/srv_pool.c [hunk 13]
```diff
@@ -4378,8 +4874,8 @@ get_open_handles_cb(daos_handle_t ih, d_iov_t *key, d_iov_t *val, void *varg)
 	/* Copy the data */
 	uuid_copy(arg->next->pic_hdl, *uuid);
 	arg->next->pic_flags = hdl->ph_flags;
-	arg->next->pic_capas = hdl->ph_sec_capas;
```
Comment (kjacque) [inline src/pool/srv_pool.c:4381]: Why is this removed?

File: src/pool/srv_pool.c [hunk 14]
```diff
@@ -3628,6 +3765,355 @@ ds_pool_prop_set_handler(crt_rpc_t *rpc)
...
+		rc = find_hdls_to_evict(tx, svc, &hdl_uuids, &hdl_uuids_size,
+					&n_hdl_uuids, NULL);
+		if (rc)
+			return rc;
+		D_DEBUG(DB_MGMT, "number of handles found was: %d\n", n_hdl_uuids);
+	}
+
+	if (n_hdl_uuids > 0) {
+		rc = pool_disconnect_hdls(tx, svc, hdl_uuids, n_hdl_uuids,
+					  rpc->cr_ctx);
```
Comment (liw) [inline src/pool/srv_pool.c:3790]: Do we also need to set ds_pool_prop_connectable to 1 before calling pool_disconnect_hdls and set it back to 0 when the upgrade terminates (successfully or unsuccessfully), so that no new connection will be established during the upgrade?

File: src/pool/srv_pool.c (hunk 12 above)
Comment (wangshilong) [reply src/pool/srv_pool.c:3942]: d_iov_set(&value, &val, sizeof(val)) is called above while updating ds_pool_prop_upgrade_status

File: src/pool/srv_pool.c [hunk 15]
```diff
@@ -3628,6 +3765,355 @@ ds_pool_prop_set_handler(crt_rpc_t *rpc)
...
+		D_GOTO(out_upgrade, rc);
+	} else {
+		d_iov_set(&value, &upgrade_status, sizeof(upgrade_status));
+		rc = rdb_tx_lookup(&tx, &svc->ps_root, &ds_pool_prop_upgrade_status,
+				   &value);
+		if (rc)
+			D_GOTO(out_tx, rc);
+
+		if (upgrade_global_ver > DS_POOL_GLOBAL_VERSION) {
+			D_ERROR("downgrading pool is unsupported\n");
```
Comment (liw) [inline src/pool/srv_pool.c:4034]: [Nit] All these error messages should be in the format 'DF_UUID": <msg>\n"' so that we know which pool has the problem and what the problem is. For instance, this one should be

    DF_UUID": downgrading pool not supported: %u -> %u\n", DP_UUID(svc->ps_uuid)

File: src/pool/srv_pool.c [hunk 16]
```diff
@@ -3628,6 +3765,355 @@ ds_pool_prop_set_handler(crt_rpc_t *rpc)
...
+	d_iov_t				value;
+	uint32_t			val;
+	int				rc1;
+	bool				need_put_leader = false;
+	daos_prop_t			*prop = NULL;
+
+	if (!svc) {
+		rc1 = pool_svc_lookup_leader(pool_uuid, &svc, NULL);
+		if (rc1 != 0)
+			return rc1;
```
Comment (liw) [inline src/pool/srv_pool.c:3920]: Do we need

    need_put_leader = true;

File: src/pool/srv_pool.c [hunk 17]
```diff
@@ -3628,6 +3765,355 @@ ds_pool_prop_set_handler(crt_rpc_t *rpc)
...
+			D_ERROR(DF_UUID": failed to upgrade global version "
+				"of pool, %d.\n", DP_UUID(pool_uuid), rc1);
+			D_GOTO(out_tx, rc1);
+		}
+	}
+
+	rc1 = pool_prop_read(&tx, svc, DAOS_PO_QUERY_PROP_ALL, &prop);
+	if (rc1)
+		D_GOTO(out_tx, rc1);
+	rc1 = ds_pool_iv_prop_update(svc->ps_pool, prop);
```
Comment (liw) [inline src/pool/srv_pool.c:3954]: Should we call this only if the rdb_tx_commit call below succeeds?

File: src/pool/srv_pool.c [hunk 18]
```diff
@@ -3628,6 +3765,355 @@ ds_pool_prop_set_handler(crt_rpc_t *rpc)
...
+		rc1 = rdb_tx_update(&tx, &svc->ps_root,
+				    &ds_pool_prop_global_version, &value);
+		if (rc) {
+			D_ERROR(DF_UUID": failed to upgrade global version "
+				"of pool, %d.\n", DP_UUID(pool_uuid), rc1);
+			D_GOTO(out_tx, rc1);
+		}
+	}
+
+	rc1 = pool_prop_read(&tx, svc, DAOS_PO_QUERY_PROP_ALL, &prop);
```
Comment (liw) [inline src/pool/srv_pool.c:3951]: Note that an RDB TX can't read its own uncommitted writes. See https://github.com/daos-stack/daos/blob/629ca9714a01d83f2dbd53a3a28de09afb81f597/src/include/daos_srv/rdb.h#L63

File: src/container/srv_container.c [hunk 19]
```diff
@@ -3440,6 +3513,168 @@ ds_cont_list(uuid_t pool_uuid, struct daos_pool_cont_info **conts,
...
+	}
+
+out:
+	if (rc == 0) {
+		ap->cont_nrs++;
+		if (upgraded) {
+			daos_prop_t *prop = NULL;
+
+			ap->cont_upgraded_nrs++;
+			rc = cont_prop_read(ap->tx, cont, DAOS_CO_QUERY_PROP_ALL, &prop);
```
Comment (liw) [inline src/container/srv_container.c:3614]: See my comment for srv_pool.c on uncommitted data visibility.

File: src/container/srv_container.c [hunk 20]
```diff
@@ -3440,6 +3513,168 @@ ds_cont_list(uuid_t pool_uuid, struct daos_pool_cont_info **conts,
...
+			if (rc)
+				goto out_put_cont;
+
+			rc = cont_iv_prop_update(ap->svc->cs_pool->sp_iv_ns,
```
Comment (liw) [inline src/container/srv_container.c:3618]: See my comment for srv_pool.c on calling this only after a successful commit.

File: src/pool/srv_pool.c [hunk 21]
```diff
@@ -4431,6 +4928,18 @@ ds_pool_get_open_handles(uuid_t pool_uuid, d_iov_t *hdls)
...
+	rc = rdb_tx_lookup(&tx, &svc->ps_root,
+			   &ds_pool_prop_upgrade_status, &value);
+	if (rc == 0 && upgrade_status == DAOS_UPGRADE_STATUS_IN_PROGRESS) {
+		D_ERROR(DF_UUID": being upgraded, not accepting connections\n",
+			DP_UUID(pool_uuid));
+		D_GOTO(out_lock, rc = -DER_BUSY);
+	} else if (rc && rc != -DER_NONEXIST) {
+		D_GOTO(out_lock, rc);
+	}
+
```
[the remark this answers is not in the document]
Comment (wangshilong) [reply src/pool/srv_pool.c:4942]: Agreed, this function shall be removed.

File: src/pool/srv_pool.c (hunk 14 above)
Comment (wangshilong) [reply src/pool/srv_pool.c:3790]: Yes, i thought i handled  this logic in the above function, i should do this.

File: src/pool/srv_pool.c (hunk 15 above)
Comment (wangshilong) [reply src/pool/srv_pool.c:4034]: Sure

File: src/pool/srv_pool.c (hunk 16 above)
Comment (wangshilong) [reply src/pool/srv_pool.c:3920]: Yes, looks we don't need such logic, caller always pass svc.

File: src/container/srv_layout.c (hunk 8 above)
Comment (wangshilong) [reply src/container/srv_layout.c:58]: I agreed this is not perfect, i might modify they share same ACL ptr.

File: src/container/srv_container.c (hunk 7 above)
Comment (wangshilong) [reply src/container/srv_container.c]: Added.

File: src/container/srv_layout.c (hunk 8 above)
Comment (kjacque) [reply src/container/srv_layout.c:58]: I wonder if maybe it would work to have a base set that represents the latest/greatest, and then have a v0 structure that includes only values that have changed. And perhaps a list of prop names we expect each version to have?

**Review by kjacque — DISMISSED** [review]

**Review by kjacque — APPROVED** [review]

**Review by liw — APPROVED** [review]

**Comment by brianjmurrell** [comment]
The checkpatch failure here looks like it was actually an error in the checkpatch processing:

```
03:39:46  2022-03-30 07:39:46,467 Creating review on try 1 got a GithubException
03:39:46  Unhandled 422 exception:
03:39:46  exception: 422 {"message": "Unprocessable Entity", "errors": ["Pull request review thread line must be part of the diff, Pull request review thread start line must be part of the same hunk as the line., and Pull request review thread diff hunk can't be blank"], "documentation_url": "https://docs.github.com/rest/reference/pulls#create-a-review-for-a-pull-request"}
03:39:46  exception.status: 422
03:39:46  exception.data: {'message': 'Unprocessable Entity', 'errors': ["Pull request review thread line must be part of the diff, Pull request review thread start line must be part of the same hunk as the line., and Pull request review thread diff hunk can't be blank"], 'documentation_url': 'https://docs.github.com/rest/reference/pulls#create-a-review-for-a-pull-request'}
```

File: src/container/srv_container.c [hunk 22]
```diff
@@ -682,6 +699,13 @@ cont_prop_write(struct rdb_tx *tx, const rdb_path_t *kvs, daos_prop_t *prop,
 			rc = rdb_tx_update(tx, kvs, &ds_cont_prop_rp_pda,
 					   &value);
 			break;
+		case DAOS_PROP_CO_GLOBAL_VERSION:
+			d_iov_set(&value, &entry->dpe_val,
```
Comment (kccain) [inline src/container/srv_container.c:703]: This writes a 64-bit number into the rdb, whereas I think the intention was for cont_global_version to be a uint32_t in the database. And the function upgrade_cont_cb() appears to be explicitly attempting to use uint32_t type for the cont_global_version value for containers that get upgraded.  I filed issue https://daosio.atlassian.net/browse/DAOS-10929 with more information.
